db/system_keyspace.cc

```diff
@@ -2403,6 +2403,23 @@ future<mutation> system_keyspace::get_group0_history(distributed<replica::databa
     co_return mutation(s, partition_key::from_singular(*s, GROUP0_HISTORY_KEY));
 }
 
+future<mutation> system_keyspace::get_group0_schema_version() {
```

Comment:
> migration_manager: store group0_schema_version in scylla_local during schema changes
>
> We extend schema mutations with an additional mutation to the
> `system.scylla_local` table which:
>  - in Raft mode, stores a UUID under the `group0_schema_version` key.
>  - outside Raft mode, stores a tombstone under that key.


Why can't we reuse the existing schema_version column?

>
> As we will see in later commits, nodes will use this after applying
> schema mutations. If the key is absent or has a tombstone, they'll
> calculate the global schema digest on their own -- using the old way. If
> the key is present, they'll take the schema version from there.

Shouldn't that be done using the new feature?

Also, we already disseminate the schema version using the schema_history table.



>
> The Raft-mode schema version is equal to the group 0 state ID of this
> schema command.
>
> The tombstone is necessary for the case of performing a schema change in
> RECOVERY mode. It will force a revert to the old digest-based way.
>


Is there a way to stop using the old digest for recovery? I want to have the ability to drop that code eventually.

> Note that extending schema mutations with a `system.scylla_local`
> mutation is possible thanks to earlier commits which moved
> `system.scylla_local` to schema commitlog, so all mutations in the
> schema mutations vector still go to the same commitlog domain.

Reply:
> Why can't we reuse the existing schema_version column?

You mean in `SCYLLA_TABLE_SCHEMA_HISTORY` table? This is a per-table schema version. In contrast, `group0_schema_version` is to store the global schema version (in old mode, a digest of all schema.)

> Shouldn't that be done using the new feature?

The absent/tombstone case is also used for schema changes done in recovery mode, while the feature is already enabled.

Besides even if we didn't have recovery mode, it would be incorrect to use the feature, because schema change propagation may race with feature enablement.

> Is there a way to stop using the old digest for recovery? I want to have the ability to drop that code eventually.

I don't know.

Reply:
> > Why can't we reuse the existing schema_version column?
> 
> You mean in `SCYLLA_TABLE_SCHEMA_HISTORY` table? This is a per-table schema version. In contrast, `group0_schema_version` is to store the global schema version (in old mode that was the digest of all schema.)


I see. I wonder what it's useful for.

> 
> > Shouldn't that be done using the new feature?
> 
> The absent/tombstone case is also used for schema changes done in recovery mode, while the feature is already enabled.
> 
> Besides even if we didn't have recovery mode, it would be incorrect to use the feature, because schema change propagation may race with feature enablement.

I don't see the connection, sorry.

> 
> > Is there a way to stop using the old digest for recovery? I want to have the ability to drop that code eventually.
> 
> I don't know.

Reply:
> I see. I wonder what it's useful for.

The global schema version?
- it's gossiped as `application_state::SCHEMA` and in non-Raft mode it was used to ensure schema convergence (nodes would pull schema if they saw that it's different in gossiper.)
- it can be queried through CQL and through REST APIs. Perhaps drivers use it for something too, but I'm not sure.

@tgrabiec -- do you recall what else the global schema version is used for?

> I don't see the connection, sorry.

Please clarify then what you mean by "using the new feature", and I'll see if there's a problem with that proposal.

I imagined you mean something like: when a node obtains a schema change and sees that the feature is enabled, it won't calculate digests but use the ones from mutations.

This has the following problem: the node which sent the schema change could have seen that the feature is enabled, so it won't calculate digests and it expects other nodes not to calculate digests as well. But the node which receives the schema change might have not yet seen that the feature is enabled, so it would incorrectly calculate digests instead of taking versions from mutations.